ncm-spma/src/main/perl/spma/yumng.pm

```diff
@@ -0,0 +1,810 @@
+    }
+    return 1;
+}
+
+=pod
+
+=head2 C<execute_yum_command>
+
+Executes C<$command> for reason C<$why>. Optionally with standard
+input C<$stdin>.  The command may be executed even under --noaction if
+C<$keeps_state> has a true value.
+
+If the command is executed, this method returns its standard output
+upon success or C<undef> in case of error.  If the command is not
+executed the method always returns a true value (usually 1, but don't
+rely on this!).
+
+Yum-based commands require annoying error handling: they may exit 0
+even upon errors.  In those cases, we have to detect errors by looking
+at stderr.  This method just encapsulates all that logic, keeping the
+callers clean.
+
+=cut
+
+my $yum_warned;
+my $yum_error;
+
+sub execute_yum_command
+{
+    my ( $self, $command, $why, $keeps_state, $stdin, $error_logger, $error_ok ) = @_;
+
+    $error_logger = "error" if ( !( $error_logger && $error_logger =~ m/^(error|warn|info|verbose)$/ ) );
+
+    $yum_warned = 0;
+
+    my ( %opts, $out, $err, @missing );
+
+    %opts = ( log => $self,
+        stdout      => \$out,
+        stderr      => \$err,
+        keeps_state => $keeps_state );
+
+    $opts{stdin} = $stdin if defined($stdin);
+
+    my $cmd = CAF::Process->new( $command, %opts );
+
+    $cmd->info("$why");
+    $cmd->execute();
+    $yum_error = $err;
+    if ( $err && ( !defined($error_ok) || $error_ok != 1 ) ) {
+        $self->warn("$why produced warnings: $err") if ( $err && ( !defined($error_ok) || $error_ok != 1 ) );
+        $yum_warned = 1;
+    }
+    $self->verbose("$why output:\n$out") if ( defined($out) );
+    if ( $? ||
+        ( $err && $err =~ m{^(?:Error|Failed|
+                      (?:Could \s+ not \s+ match)|
+                      (?:Transaction \s+.*rror)|
+                      (?:Unknown \s+ group \s+  package \s+ type) |
+                      (?:.*requested \s+ URL \s+ returned \s+ error))}oxmi ) ||
+        ( $out && ( @missing = ( $out =~ m{^No package (.*) available}omg ) ) )
+      ) {
+        return undef if ( !defined($error_ok) || $error_ok != 1 );
+        $self->$error_logger( "Failed $why: ", $err || "(empty/undef stderr)" );
+        if (@missing) {
+            $self->$error_logger( "Missing packages: ", join ( " ", @missing ) );
+        }
+        return $err;
+    }
+    return $cmd->{NoAction} || $out;
+}
+
+sub execute_yum_command_with_recovery
+{
+    my ( $self, $command, $why, $keeps_state, $stdin, $error_logger, $error_ok ) = @_;
+    my $status = $self->execute_yum_command($command, $why, $keeps_state, $stdin, $error_logger, $error_ok);
+    my $packages_modified = 0;
+
+    if ( !defined($status) ) {
+	my @lines = split /\n/, $yum_error;
+	foreach my $line (@lines) {
+	    my $cf = index($line, 'conflicts with file from package');
+	    if ( $cf != -1) {
+		my $pkg = $line;
+		$pkg =~ s,^.*package ,,;
+		$pkg =~ s,-[0-9]*:,-,;
+		$self->execute_yum_command(["rpm -e --nodeps $pkg"], "removing $pkg to avoid file conflict");
```

Comment:
you might want to document what this is trying to achieve. (because this is not forced, not too many teriible things can happen, but on the otehr hand, removal of packages with dependencies between them might fail if removed in wrong order).